### ISSUE TITLE
assign operator may suppress the next print

### ISSUE THREAD

**Issue by shrektan** [Rdatatable/data.table#4331]
I haven't looked into the cause of this issue yet but I think this should be improved. Despite I'm aware of the fix to this issue is to **change the last `x` to `x[]`**, many users don't. They thought the function returns nothing at first glance and would be very very confused.

``` r
library(data.table)
fun <- function() {
  x <- data.table(A = 1:3)
  x[, B := 3]
  x
}
dt <- fun()
dt # suppressed
dt # has to call it twice
#>    A B
#> 1: 1 3
#> 2: 2 3
#> 3: 3 3
```

<sup>Created on 2020-03-31 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by nbenn** [comment]
I'm sure others have spent more time thinking about this and I only had a quick look into how printing is suppressed currently. It seems to me that

```r
x <- data.table(A = 1:3)
x[, B := 3]
x
#>    A B
#> 1: 1 3
#> 2: 2 3
#> 3: 3 3
```

works as expected, b/c printing is triggered when running `x[, B := 3]` and therefore print suppression is reset, whereas when, as is the case in your example, `x[, B := 3]` is not a top-level expression and therefore print suppression is not reset, leading to down-stream issues. @shrektan Wouldn't this problem go away if `[.data.table()` would check whether called on top-level and only then suppressing printing? If not called on top-level, printing does not happen anyways.

**Comment by shrektan** [comment]
TL;DR

Yes, this was my initial idea. However, this won't work for cases like `knitr`, where the code is evaluated differently. In other words, unless we can find a way to reliably change the visibility *(see `?withVisible`)* of the returned object *(like adding a hook function to `[` but I'm not sure if there's a way to do that in R)* to avoid the `printing` from the root, we can only do the **hacking** stuff in `print.data.table()`. 

About the **hacking** stuff, my idea is to check not only the object's memory address, but also the relationships of the calling environments as well. I did some proto-codes recently. It seems promising. However, there're two issues need to be resolved:

1. "storing" the calling environment may affect the performance of data.table and how can we minimize it?
1. We need to hack different for at least two cases: interactive use, knitr, as they are quite different... It makes things complicate so that we can't be sure there's no unexpected side-effects introduced. 

**In short, I think we can improve this by hacking but personally I'm not very satisfied this solution.** 

The current hacking is not perfect but the logic is simple and clear. At least, you know it has minimal impact on performance, you know when the hack won't work and it has been tested for years...
